proposals/imporved-hash-module.md

```diff
@@ -0,0 +1,134 @@
+# Improve hash module
+
+Current implementation of the hash module in standard library reflex the implementation of the Python hash module, which by itself is a good idea, but it has some flaws, which we should correct in Mojo.
+
+# Flaws of the `Hashable` trait
+
+The `Hashable` trait is designed as following:
+
+```mojo
+trait Hashable:
+    fn __hash__(self) -> Int:
+        ...
+```
+
+Which implies that a developer, who writes a new hashable struct needs to return an `Int`. 
+- This API does not provide guidance to how this `Int` value needs to be computed
+- It is impossible to exchange hashing algorithms on demand
+- `Int` as a type is variable length (based on the CPU arch), which might cause issues
+- Such trait design follows the call return principle which, when applied on complex types, will lead to unnecessary computations and memory allocations
+
+## Example
+
+```mojo
+@value
+struct Person(Hashable):
+    var name: String
+    var age: UInt8
+    var friends_names: List[String]
+
+    fn __hash__(self) -> Int:
+        var hashes = List[Int]()
+        hashes.append(hash(self.name))
+        hashes.append(hash(self.age))
+        for friend in self.friends_names:
+            hashes.append(hash(friend[]))
+        
+        # How to combine a hash of hashes ???
+```
+As you can see above we, computed hashes for all of the struct fields, but we are uncertain how to combine those values in a way which produces a good (non compromised) hash value.
+
+# Proposal
+
+In order to improve the hash module and address the flaws of the `Hashable` trait, we need to apply following steps.
+
+## Introduce a `Hasher` trait
+
+By introducing a `Hasher` trait we define an abstraction for the hashing algorithm itself, which allows streaming creation of the hash value. Bellow is a possible API of the `Hashable` trait.
+
+```mojo
+trait Hasher:
+    fn __init__(inout self):
+        ...
+    fn update[T: DType](inout self, value: SIMD[T, 1]):
+        ...
+    fn update(inout self, pointer: DTypePointer[DType.uint8], length: Int):
+        ...
+    fn finish(owned self) -> UInt64:
+        ...
+```
+
+## Implement a default `Hasher` in standard library
+
+The standard library should provide a default `Hasher` implementation, but it would be possible for the developers to implement, or choose other hash algorithms, if they better fit their use case.
+
+Bellow you can see a dummy implementation of a `DefaultHasher`
+
+```mojo
+struct DefaultHasher(Hasher):
+    var hash: UInt64
+
+    fn __init__(inout self):
+        self.hash = 42
+    fn update[T: DType](inout self, value: SIMD[T, 1]):
+        ...
+    fn update(inout self, pointer: DTypePointer[DType.uint8], length: Int):
+        ...
```

Comment:
Can we add a comment saying that when Mojo supports it, we could have a default implementation for this?
Something like:
```suggestion
    fn update(inout self, pointer: DTypePointer[DType.uint8], length: Int):
        for i in range(length):
            self.update(pointer[i])
```
Then a user can override the method to implement something more efficient if they wish to do so.

Reply:
See above, I think it is a detail not really important for this proposal.

Reply:
It was just a suggestion, fair enough